openpathsampling/experimental/storage/sql_backend.py

```diff
@@ -0,0 +1,449 @@
+        self.schema = {}
+        self.table_to_number = {}
+        self.number_to_table = {}
+        self.engine = engine
+        self.mode = mode
+        self._initialize_with_mode(self.mode)
+
+
+    def close(self):
+        # is this necessary?
+        self.engine.dispose()
+
+    @property
+    def metadata(self):
+        return self._metadata
+
+    @staticmethod
+    def filename_from_dialect(filename, dialect):
+        # TODO: I think this might be removed in the future; instead, custom
+        # setups can use the from_engine constructor
+
+        # take dialects like "sqlite", etc and return proper file connection
+        # URI; would be essentially no-op for regular file opening as with
+        # .nc
+        uri_root = {
+            'sqlite': "sqlite:///{filename}",
+        }[dialect]
+        return uri_root.format(filename=filename)
+
+    def internal_tables_from_db(self):
+        """Obtain mappings of table name to number from database.
+        """
+        tables = self.metadata.tables['tables']
+        with self.engine.connect() as conn:
+            res = conn.execute(tables.select())
+            table_to_number = {r.name: r.idx for r in res}
+        number_to_table = {v: k for (k, v) in table_to_number.items()}
+        return table_to_number, number_to_table
+
+    def _add_table_to_tables_list(self, table_name, table_schema, cls_):
+        if table_name in ['uuid', 'tables']:
+            return
+        # note that this return the number of tables in 'tables', which does
+        # not include 'uuid' or 'tables'
+        # There must be a better way to do this, but this seems to get the
+        # job done,
+        tables = self.metadata.tables['tables']
+        with self.engine.connect() as conn:
+            res = conn.execute(tables.select())
+            n_tables = len(list(res))
+
+        schema_table = self.metadata.tables['schema']
+
+        module = cls_.__module__
+        class_name = cls_.__name__
+
+        with self.engine.connect() as conn:
+            conn.execute(tables.insert().values(name=table_name,
+                                                idx=n_tables,
+                                                module=module,
+                                                class_name=class_name))
+            conn.execute(schema_table.insert().values(
+                table=table_name,
+                schema=json.dumps(table_schema)
+            ))
+
+        self.table_to_number.update({table_name: n_tables})
+        self.number_to_table.update({n_tables: table_name})
+
+    def _load_from_table(self, table_name, idx_list):
+        # this is not public API (assumes idx_list, which is reserved by not
+        # guaranteed)
+        table = self.metadata.tables[table_name]
+        or_stmt = sql.or_(*(table.c.idx == idx for idx in idx_list))
+        sel = table.select(or_stmt)
+        with self.engine.connect() as conn:
+            results = list(conn.execute(sel))
```

Comment:
So, just commenting the fix that I had to get around the hard capped SQLite limit of 10000 expressions in the expression tree ([that can only be changed at compile time](https://www.sqlite.org/limits.html)):
1. `from itertools import zip_longest` should be added to the imports

2. 
```suggestion
        # Loop over max 1000 items at once
        # taken from https://stackoverflow.com/a/434411
        results = []
        max_items = 1000
        iter_list = [iter(idx_list)] * max_items
        full_list = zip_longest(*iter_list)
        with self.engine.connect() as conn:
            for sub_list in full_list:
                or_stmt = sql.or_(*(table.c.idx == idx for idx in sub_list
                                               if idx is not None))
                sel = table.select(or_stmt)
                sub_results = list(conn.execute(sel))
                results.extend(sub_results)
```

Reply:
I took the basic idea, but details of the implementation were to have a reusable tool in the `tools.py`. Changes in 1613b34.